analyzer/src/main/kotlin/managers/utils/DependencyGraphBuilder.kt

```diff
@@ -0,0 +1,324 @@
+
+    /** The mapping from scopes to dependencies constructed by this builder. */
+    private val scopeMapping = mutableMapOf<String, List<RootDependencyIndex>>()
+
+    /** Stores all packages encountered in the dependency tree. */
+    private val resolvedPackages = mutableSetOf<Package>()
+
+    /**
+     * A set storing the packages that are direct dependencies of one of the scopes. These are the entry points into
+     * the dependency graph.
+     */
+    private val directDependencies = mutableSetOf<DependencyReference>()
+
+    /**
+     * Add the scope with the given [scopeName] to this builder. In most cases, it is not necessary to add scopes
+     * explicitly, as they are recorded automatically by _addDependency()_. However, if there are scopes without
+     * dependencies, this function can be used to include them into the builder result.
+     */
+    fun addScope(scopeName: String) {
+        scopeMapping.putIfAbsent(scopeName, emptyList())
+    }
+
+    /**
+     * Add the given [dependency] for the scope with the given [scopeName] to this builder. This function needs to be
+     * called all the dependencies of all scopes. That way the builder gets sufficient information to construct the
+     * [DependencyGraph].
+     */
+    fun addDependency(scopeName: String, dependency: D) {
+        addDependencyToGraph(scopeName, dependency, transitive = false)
+    }
+
+    /**
+     * Construct the [DependencyGraph] from the dependencies passed to this builder so far.
+     */
+    fun build(): DependencyGraph = DependencyGraph(dependencyIds, directDependencies, scopeMapping)
+
+    /**
+     * Return a set with all the packages that have been encountered for the current project.
+     */
+    fun packages(): Set<Package> = resolvedPackages
+
+    /**
+     * Update the dependency graph by adding the given [dependency], which may be [transitive], for the scope with name
+     * [scopeName]. All the dependencies of this dependency are processed recursively.
+     */
+    private fun addDependencyToGraph(scopeName: String, dependency: D, transitive: Boolean): DependencyReference {
+        val identifier = dependencyHandler.identifierFor(dependency)
+        val issues = dependencyHandler.issuesForDependency(dependency).toMutableList()
+        val index = updateDependencyMappingAndPackages(identifier, dependency, issues)
+
+        val ref = when (val result = findDependencyInGraph(index, dependency)) {
+            is GraphSearchResult.Found -> result.ref
+            is GraphSearchResult.NotFound ->
+                insertIntoGraph(
+                    RootDependencyIndex(index, result.fragmentIndex),
+                    scopeName,
+                    dependency,
+                    issues,
+                    transitive
+                )
+            is GraphSearchResult.Incompatible ->
+                insertIntoNewFragment(index, scopeName, dependency, issues, transitive)
```

Comment:
Formatting: Code for a case should be wrapped in `{}` if it doesn't fit in one line, consider adding empty lines between the cases (https://kotlinlang.org/docs/coding-conventions.html#control-flow-statements).